File: apps/api/app/core/phone.py
"""Tanzania mobile number normalization.

Accepts any of the three formats a human or a MikroTik hotspot form is
likely to produce and normalizes them all to the same canonical
`255XXXXXXXXX` (12 digits, no leading `+`) used everywhere internally:

    07XXXXXXXX      (10 digits, local)
    +2557XXXXXXXX   (international, with +)
    2557XXXXXXXX    (international, no +)

The subscriber number itself is validated only against Tanzania's national
numbering plan shape (9 digits, starting 6 or 7) — deliberately not
restricted to any single operator's prefix range, since prefixes are
reassigned and ported between operators over time.
"""

import re

_LOCAL = re.compile(r"^0([67]\d{8})$")
_INTERNATIONAL = re.compile(r"^\+?255([67]\d{8})$")


def normalize_tz_phone(raw: str) -> str:
    """Returns the canonical `255XXXXXXXXX` form, or raises ValueError."""
    cleaned = re.sub(r"[\s\-()]", "", raw.strip())

    if match := _LOCAL.match(cleaned):
        return f"255{match.group(1)}"
    if match := _INTERNATIONAL.match(cleaned):
        return f"255{match.group(1)}"

    raise ValueError(
        f"Invalid Tanzania phone number: {raw!r} "
        "(expected 07XXXXXXXX, +2557XXXXXXXX, or 2557XXXXXXXX)"
    )


def is_valid_tz_phone(raw: str) -> bool:
    try:
        normalize_tz_phone(raw)
    except ValueError:
        return False
    return True

File: apps/web/lib/collections/phone.ts
/**
 * Tanzania mobile number normalization for the Collection form.
 *
 * Deliberately mirrors apps/api/app/core/phone.py exactly — same two
 * patterns, same canonical `255XXXXXXXXX` output, same national-numbering
 * shape (9 digits starting 6 or 7, not restricted to a single operator's
 * prefix range since prefixes get ported).
 *
 * This is a UX convenience only: it catches an obvious typo before an STK
 * request is spent on it. The backend re-normalizes and re-validates every
 * number it receives and stays the authority — this never relaxes anything
 * the server would reject.
 */

const LOCAL = /^0([67]\d{8})$/;
const INTERNATIONAL = /^\+?255([67]\d{8})$/;

export type PhoneNormalizationResult =
  | { ok: true; value: string }
  | { ok: false; reason: string };

/** Strips the separators a human might type: spaces, dashes, parentheses. */
function clean(raw: string): string {
  return raw.trim().replace(/[\s\-()]/g, "");
}

/**
 * Normalizes `0762474101`, `+255762474101` or `255762474101` to
 * `255762474101`. Returns a reason instead of throwing so the form can
 * render it inline.
 */
export function normalizeTzPhone(raw: string): PhoneNormalizationResult {
  const cleaned = clean(raw);

  if (cleaned === "") {
    return { ok: false, reason: "Enter the customer's phone number." };
  }

  const local = LOCAL.exec(cleaned);
  if (local) return { ok: true, value: `255${local[1]}` };

  const international = INTERNATIONAL.exec(cleaned);
  if (international) return { ok: true, value: `255${international[1]}` };

  return {
    ok: false,
    reason: "Enter a valid Tanzanian mobile number, e.g. 0762474101 or +255762474101.",
  };
}

export function isValidTzPhone(raw: string): boolean {
  return normalizeTzPhone(raw).ok;
}

/**
 * Masks a normalized number for display where the full number isn't
 * needed. Matches the backend's own masking shape (see
 * apps/api/app/schemas/finance.py._mask_phone) so a masked value coming
 * from the API and one rendered locally look identical.
 */
export function maskTzPhone(phone: string): string {
  if (phone.length <= 7) return phone;
  return `${phone.slice(0, 4)}${"*".repeat(phone.length - 7)}${phone.slice(-3)}`;
}
